keps/sig-node/20190717-seccomp-ga.md

```diff
@@ -0,0 +1,361 @@
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This is a proposal to upgrade the seccomp annotation on pods & pod security policies to a field, and
+mark the feature as GA. This proposal aims to do the _bare minimum_ to clean up the feature, without
+blocking future enhancements.
+
+## Motivation
+
+Docker started enforcing a default seccomp profile in v1.10. At the time, Kubernetes (in v1.2)
+didn't have a way to control the seccomp profile, so the profile was disabled (set to `unconfined`)
+to prevent a regression (see https://github.com/kubernetes/kubernetes/pull/21790). In Kubernetes
+v1.3, annotations were added to give users some control over the profile:
+
+```
+seccomp.security.alpha.kubernetes.io/pod: {unconfined,docker/default,localhost/<path>}
+container.seccomp.security.alpha.kubernetes.io/<container_name>: ...
+```
+
+The feature has been more or less unchanged ever since. Also note that the addition predates feature
+gates or our modern concept of feature lifecycle. So, even though the annotations include `alpha` in
+the key, this is entirely useable on any production GA cluster.
+
+There have been multiple attempts to [change the default
+profile](https://github.com/kubernetes/kubernetes/issues/39845) or [formally spec the Kubernetes
+seccomp profile](https://github.com/kubernetes/kubernetes/issues/39128), but both efforts were
+abandoned due to friction and lack of investment.
+
+Despite the `alpha` label, I think this feature needs to be treated as GA, and we're doing our users
+a disservice by leaving it in this weird limbo state. As much as I would like to see seccomp support
+fully fleshed out, if we block GA on those enhancements we will remain stuck in the current state
+indefinitely. Therefore, I'm proposing we do the absolute minimum to clean up the current
+implementation all accurately declare the feature "GA". Future enhancements can follow the standard
+alpha -> beta -> GA feature process.
+
+_NOTE: AppArmor is in a very similar state, but with some subtle differences. Promoting AppArmor to
+GA will be covered by a separate KEP._
+
+### Goals
+
+- Declare seccomp GA
+- Fully document and formally spec the feature support
+- Migrate the annotations to standard API fields
```

Comment:
suggest using language like "add equivalent API fields" rather than "migrate" (which sounds like we will move data, which we know doesn't work well)

Reply:
If the pod fields are immutable, is it possible to still set them on update as part of the storage flow? The criteria would be:

1. The seccomp annotations are set, and do not change as part of the update.
2. The seccomp fields are not set.

I guess the first criteria is problematic, since the seccomp fields could be changed in the update, and then a subsequent update could leave them be, so maybe that doesn't matter.